pkg/authorization/registry/role/policybased/virtual_storage.go

```diff
@@ -0,0 +1,198 @@
+func (m *VirtualStorage) List(ctx kapi.Context, label labels.Selector, field fields.Selector) (runtime.Object, error) {
+	policyList, err := m.policyStorage.ListPolicies(ctx, labels.Everything(), fields.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	roleList := &authorizationapi.RoleList{}
+
+	for _, policy := range policyList.Items {
+		for _, role := range policy.Roles {
+			if label.Matches(labels.Set(role.Labels)) {
+				roleList.Items = append(roleList.Items, role)
+			}
+		}
+	}
+
+	return roleList, nil
+}
+
+func (m *VirtualStorage) Get(ctx kapi.Context, name string) (runtime.Object, error) {
+	policy, err := m.policyStorage.GetPolicy(ctx, authorizationapi.PolicyName)
+	if err != nil && kapierrors.IsNotFound(err) {
+		return nil, kapierrors.NewNotFound("Role", name)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	role, exists := policy.Roles[name]
+	if !exists {
+		return nil, kapierrors.NewNotFound("Role", name)
+	}
+
+	return &role, nil
+}
+
+// Delete(ctx api.Context, name string) (runtime.Object, error)
+func (m *VirtualStorage) Delete(ctx kapi.Context, name string, options *kapi.DeleteOptions) (runtime.Object, error) {
+	policy, err := m.policyStorage.GetPolicy(ctx, authorizationapi.PolicyName)
+	if err != nil && kapierrors.IsNotFound(err) {
+		return nil, kapierrors.NewNotFound("Role", name)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	if _, exists := policy.Roles[name]; !exists {
+		return nil, kapierrors.NewNotFound("Role", name)
+	}
+
+	delete(policy.Roles, name)
+	policy.LastModified = util.Now()
+
+	if err := m.policyStorage.UpdatePolicy(ctx, policy); err != nil {
+		return nil, err
+	}
+	return &kapi.Status{Status: kapi.StatusSuccess}, nil
+}
+
+func (m *VirtualStorage) Create(ctx kapi.Context, obj runtime.Object) (runtime.Object, error) {
+	if err := rest.BeforeCreate(roleregistry.Strategy, ctx, obj); err != nil {
+		return nil, err
+	}
+
+	role := obj.(*authorizationapi.Role)
+
+	policy, err := m.EnsurePolicy(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if _, exists := policy.Roles[role.Name]; exists {
+		return nil, kapierrors.NewAlreadyExists("Role", role.Name)
+	}
+
+	role.ResourceVersion = policy.ResourceVersion
+	policy.Roles[role.Name] = *role
+	policy.LastModified = util.Now()
+
+	if err := m.policyStorage.UpdatePolicy(ctx, policy); err != nil {
+		return nil, err
+	}
+
+	return role, nil
+}
+
+func (m *VirtualStorage) Update(ctx kapi.Context, obj runtime.Object) (runtime.Object, bool, error) {
+	role, ok := obj.(*authorizationapi.Role)
+	if !ok {
+		return nil, false, kapierrors.NewBadRequest(fmt.Sprintf("obj is not a role: %#v", obj))
+	}
+
+	old, err := m.Get(ctx, role.Name)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if err := rest.BeforeUpdate(roleregistry.Strategy, ctx, obj, old); err != nil {
+		return nil, false, err
+	}
+
+	policy, err := m.policyStorage.GetPolicy(ctx, authorizationapi.PolicyName)
+	if err != nil && kapierrors.IsNotFound(err) {
```

Comment:
same comment here, but guess since this returns three arguments, maybe not the same function block.